library/bignum_mod.h

```diff
@@ -0,0 +1,142 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#ifndef MBEDTLS_BIGNUM_MOD_H
+#define MBEDTLS_BIGNUM_MOD_H
+
+#include "common.h"
+
+#if defined(MBEDTLS_BIGNUM_C)
+#include "mbedtls/bignum.h"
+#endif
+
+
+typedef struct
+{
+    size_t n;
+    mbedtls_mpi_uint *p;
+} mbedtls_mpi_mod_residue;
+
+typedef void* mbedtls_mpi_mont_struct;
+typedef void* mbedtls_mpi_opt_red_struct;
+
+typedef struct {
+    mbedtls_mpi_uint *p;
+    size_t n;       // number of limbs
+    size_t plen;    // bitlen of p
+    int ext_rep;    // signals external representation (eg. byte order)
+    int int_rep;    // selector to signal the active member of the union
+    union rep
+    {
+        mbedtls_mpi_mont_struct mont;
+        mbedtls_mpi_opt_red_struct ored;
+    } rep;
+} mbedtls_mpi_mod_modulus;
+
+typedef enum
+{
+    MBEDTLS_MPI_MOD_REP_INVALID    = 0,
+    MBEDTLS_MPI_MOD_REP_MONTGOMERY,
+    MBEDTLS_MPI_MOD_REP_OPT_RED
+} mbedtls_mpi_mod_rep_selector;
+
+typedef enum
+{
+    MBEDTLS_MPI_MOD_EXT_REP_INVALID    = 0,
+    MBEDTLS_MPI_MOD_EXT_REP_LE,
+    MBEDTLS_MPI_MOD_EXT_REP_BE
+} mbedtls_mpi_mod_ext_rep;
+
+
+/** Setup a residue structure.
+ *
+ * \param r     The address of residue to setup. The size is determined by \p m.
+ *              (In particular, it must have at least as many limbs as the
+ *              modulus \p m.)
+ * \param m     The address of the modulus related to \p r.
+ * \param p     The address of the limb array storing the value of \p r. The
+ *              memory pointed by \p p will be used by \p r and must not be
+ *              freed or written until after mbedtls_mpi_mod_residue_release()
+ *              is called.
+ * \param pn    The number of limbs of \p p.
+ *
+ * \return      \c 0 if successful.
+ * \return      #MBEDTLS_ERR_MPI_BAD_INPUT_DATA if \p r, \p m or \p p is #NULL
+ *              pointer, \p pn is less than the limbs in \p m or if \p p is not
+ *              less than \p m.
+ */
+int mbedtls_mpi_mod_residue_setup( mbedtls_mpi_mod_residue *r,
+                                   mbedtls_mpi_mod_modulus *m,
+                                   mbedtls_mpi_uint *p,
+                                   size_t pn );
+
+/** Unbind elements of a residue structure.
+ *
+ * This function removes the reference to the limb array that was passed to
+ * mbedtls_mpi_mod_residue_setup() to make it safe to free or use again.
+ *
+ * This function invalidates \p r and it must not be used until after
+ * mbedtls_mpi_mod_residue_setup() is called on it again.
+ *
+ * \param r     The address of residue to release.
+ */
+void mbedtls_mpi_mod_residue_release( mbedtls_mpi_mod_residue *r );
+
+/** Initialize a modulus structure.
+ *
+ * \param m     The address of a modulus.
+ */
+void mbedtls_mpi_mod_modulus_init( mbedtls_mpi_mod_modulus *m );
+
+/** Setup a residue structure.
+ *
+ * \param m         The address of a modulus.
```

Comment:
Rather than "a" random modulus, isn't this:
```suggestion
 * \param m         The address of the modulus structure to populate.
```